src/transformers/modeling_attn_mask_utils.py

```diff
@@ -316,7 +316,12 @@ def _prepare_4d_causal_attention_mask_for_sdpa(
         or (hasattr(torch, "_dynamo") and torch._dynamo.is_compiling())
     )
 
-    if attention_mask is not None:
+    ignore_causal_mask = False
+
+    if attention_mask is None:
+        if sliding_window is None or key_value_length < sliding_window:
+            ignore_causal_mask = not is_tracing
+    elif sliding_window is None or key_value_length < sliding_window:
```

Comment:
There are basically 2 cases:
1. You ignore the causal mask 
2. You don't ignore it. 
Code is really not super clear but we will refactor this soon anyways.